Frontend/src/pages/registration_edit/components/RegistrationEditor.jsx

```diff
@@ -1,35 +1,43 @@
 import { EventSelector } from '@thewca/wca-components'
-import React, { useEffect, useState } from 'react'
+import React, { useContext, useEffect, useState } from 'react'
 import { Button, Checkbox, TextArea } from 'semantic-ui-react'
-import { useHeldEvents } from '../../../api/helper/hooks/use_competition_info'
-import useCompetitorInfo from '../../../api/helper/hooks/use_competitor_info'
+import { CompetitionContext } from '../../../api/helper/context/competition_context'
 import { getSingleRegistration } from '../../../api/registration/get/get_registrations'
 import { updateRegistration } from '../../../api/registration/patch/update_registration'
 import { setMessage } from '../../../ui/events/messages'
 import LoadingMessage from '../../../ui/messages/loadingMessage'
 import styles from './editor.module.scss'
+import { useQuery } from '@tanstack/react-query'
+import getCompetitorInfo from '../../../api/user/get/get_user_info'
 
 export default function RegistrationEditor({ user_id, competition_id }) {
-  const { isLoading: eventsLoading, heldEvents } = useHeldEvents(competition_id)
-  const { isLoading: infoLoading, competitorInfo } = useCompetitorInfo(user_id)
-  const [registration, setRegistration] = useState({})
+  const { competitionInfo } = useContext(CompetitionContext)
   const [comment, setComment] = useState('')
   const [status, setStatus] = useState('')
   const [selectedEvents, setSelectedEvents] = useState([])
+  const [registration, setRegistration] = useState({})
+  const { data: serverRegistration } = useQuery({
+    queryKey: [competition_id, user_id],
+    queryFn: () => getSingleRegistration(user_id, competition_id),
+  })
+  const { isLoading, data: competitorInfo } = useQuery({
+    queryKey: ['info', user_id],
+    queryFn: () => getCompetitorInfo(user_id),
+    enabled: Boolean(serverRegistration),
+  })
 
   useEffect(() => {
```

Comment:
`useQuery` takes an `onSuccess` param in the object (and `onError` etc), which would be the cleanest place to put this.
Although maybe that would be triggered every time the query refetches? I forget. If so, then that wouldn't work.
But either way, if the query silently refetches (due to periodic invalidation of cache, or if the page is refocused) and the data has changed in the backend, then this effect is going to set all the state again, overwriting any edits the user was currently making.

Generally, when a query is being used to populate a form like this, my preference is to pass in some params to `useQuery` which prevent it from re-running after the first fetch, and then set all the state in the `onSuccess` param. I can write that out if you want an example.

Reply:
yes, please.

Reply:
Edit: Just realized that the documentation says `onSuccess` is being deprecated in the next major version, so maybe I need to do some more investigating. There might be a better/more standard way. But this is what I was going to suggest:

```
const { ... } = useQuery({
  queryKey: ...,
  queryFn: ...,
  refetchOnWindowFocus: false,
  refetchOnReconnect: false,
  staleTime: Infinity,
  refetchOnMount: 'always',
  onSuccess: populateInputs,
  onError: ...,
});

function populateInputs(serverRegistration) {
  setRegistration(serverRegistration.registration);
  ... // the other 3
}
```
So the data fetches when the page is mounted, but on no other occasions.